core/src/main/java/io/questdb/griffin/engine/groupby/GroupByAllocator.java

```diff
@@ -0,0 +1,213 @@
+        for (int i = 0, n = arenas.size(); i < n; i++) {
+            allocated += arenas.getQuick(i).allocated();
+        }
+        return allocated;
+    }
+
+    /**
+     * This method is not thread-safe and shouldn't be called concurrently
+     * with any alloc/free calls.
+     */
+    @Override
+    public void close() {
+        for (int i = 0, n = arenas.size(); i < n; i++) {
+            arenas.getQuick(i).close();
+        }
+    }
+
+    /**
+     * Best-effort free memory operation. The memory shouldn't be used after it was called.
+     */
+    public void free(long ptr, long size) {
+        tlArena.get().free(ptr, size);
+    }
+
+    public long malloc(long size) {
+        return tlArena.get().malloc(size);
+    }
+
+    public long realloc(long ptr, long oldSize, long newSize) {
+        return tlArena.get().realloc(ptr, oldSize, newSize);
+    }
+
+    private Arena newArena() {
+        final Arena arena = new Arena();
+        synchronized (lock) {
+            arenas.add(arena);
+        }
+        return arena;
+    }
+
+    private class Arena implements QuietCloseable {
+        // Holds <ptr, size> pairs.
+        private final LongLongHashMap chunks = new LongLongHashMap();
+        private long allocated;
+        private long lim;
+        private long ptr;
+
+        // Allocated chunks total (bytes).
+        public long allocated() {
+            return allocated;
+        }
+
+        @Override
+        public void close() {
+            for (int i = 0, n = chunks.capacity(); i < n; i++) {
+                long ptr = chunks.keyAtRaw(i);
+                if (ptr != -1) {
+                    long size = chunks.valueAtRaw(i);
+                    Unsafe.free(ptr, size, MemoryTag.NATIVE_GROUP_BY_FUNCTION);
+                }
+            }
+            chunks.restoreInitialCapacity();
+            allocated = 0;
+            ptr = lim = 0;
+        }
+
+        // Best-effort free operation.
+        public void free(long ptr, long size) {
+            if (size < defaultChunkSize) {
+                // We don't free small allocations.
+                return;
+            }
+            int index = chunks.keyIndex(ptr);
+            if (index < 0) {
+                long chunkSize = chunks.valueAt(index);
+                if (size == chunkSize) {
+                    // We're lucky! We can free the whole chunk.
+                    Unsafe.free(ptr, chunkSize, MemoryTag.NATIVE_GROUP_BY_FUNCTION);
+                    chunks.removeAt(index);
+                    allocated -= chunkSize;
+                    if (this.ptr == Bytes.align8b(ptr + chunkSize)) {
+                        this.ptr = lim = 0;
+                    }
+                }
+            }
+        }
+
+        public long malloc(long size) {
+            if (size > maxChunkSize) {
+                throw CairoException.nonCritical().put("too large allocation requested: ").put(size);
+            }
+
+            if (ptr + size <= lim) {
```

Comment:
```suggestion
            if (Bytes.align8b(ptr + size) <= lim) {
```

Reply:
That's not required. It's ok if `ptr` goes beyond `lim` once we align it - that's because we always check `ptr` against `lim` in `malloc`/`realloc`.